### ISSUE TITLE
`readdir_r` is deprecated in glibc 2.24

### ISSUE THREAD

**Issue by jaen** [dotnet/corefx#10712]
Hello,

The `pal_io.cpp` file in `System.Native` is using a newly deprecated function `readdir_r` which makes the sources not compile for me, with the following error message:

```
==> Starting build()...
Detected Linux x86_64
/home/jaen/.cache/pacaur/dotnet-cli/src/corefx-1.0.0/src/Native/System.Native/pal_io.cpp:344:17: error: 'readdir_r' is
      deprecated [-Werror,-Wdeprecated-declarations]
    int error = readdir_r(dir, entry, &result);
                ^
/usr/include/dirent.h:183:12: note: 'readdir_r' has been explicitly marked deprecated here
extern int readdir_r (DIR *__restrict __dirp,
           ^
1 error generated.
make[2]: *** [System.Native/CMakeFiles/System.Native-Static.dir/build.make:135: System.Native/CMakeFiles/System.Native-Static.dir/pal_io.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
/home/jaen/.cache/pacaur/dotnet-cli/src/corefx-1.0.0/src/Native/System.Native/pal_io.cpp:344:17: error: 'readdir_r' is
      deprecated [-Werror,-Wdeprecated-declarations]
    int error = readdir_r(dir, entry, &result);
                ^
/usr/include/dirent.h:183:12: note: 'readdir_r' has been explicitly marked deprecated here
extern int readdir_r (DIR *__restrict __dirp,
           ^
1 error generated.
make[2]: *** [System.Native/CMakeFiles/System.Native.dir/build.make:135: System.Native/CMakeFiles/System.Native.dir/pal_io.cpp.o] Error 1
```

Are there any plans to update this file for it to build with new glibc?

**Comment by hak8or** [comment]
Came to chime in and say that I am also hit by this. [Here](http://lwn.net/Articles/696474/) is an LWN discussion about readdir_r and why it's deprecated.

**Comment by LaurentTreguier** [comment]
It doesn't look too hard to fix, and it's only ever used in that one file. I've tried to patch it myself and successfully compiled it, but as I have no clue about how the code really works, I don't know it works as intended

**Comment by joshfree** [comment]
cc @stephentoub

**Comment by sixpindin** [comment]
@hak8or : What was your patch? - I'm trying myself in the meantime. Or if anybody else knows of a patch for this please let me know.

**Comment by hak8or** [comment]
@sixpindin You might be getting me mixed up with @LaurentTreguier , he was the one who made the patch, not me.

**Comment by LaurentTreguier** [comment]
@sixpindin I simply tried to replace `readdir_r` with `readdir`... However I have no idea of how the function is really used and this is really a random try. `readdir` uses one parameter less than `readdir_r` (the `entry` parameter), which seems to be used later in the code, so I don't really think this is a good solution.
I replaced this :
```cpp
    dirent* result = nullptr;
    dirent* entry = static_cast<dirent*>(buffer);
    int error = readdir_r(dir, entry, &result);

    // positive error number returned -> failure
    if (error != 0)
    {
        assert(error > 0);
        *outputEntry = {}; // managed out param must be initialized
        return error;
    }

    // 0 returned with null result -> end-of-stream
    if (result == nullptr)
    {
        *outputEntry = {}; // managed out param must be initialized
        return -1;         // shim convention for end-of-stream
    }

    // 0 returned with non-null result (guaranteed to be set to entry arg) -> success
    assert(result == entry);
    ConvertDirent(*entry, outputEntry);
    return 0;
```
with this :
```cpp
    dirent* result = nullptr;
    result = readdir(dir);

    // positive error number returned -> failure
    if (result != nullptr)
    {
        *outputEntry = {}; // managed out param must be initialized
        return result != nullptr;
    }

    // 0 returned with null result -> end-of-stream
    if (result == nullptr)
    {
        *outputEntry = {}; // managed out param must be initialized
        return -1;         // shim convention for end-of-stream
    }

    ConvertDirent(*entry, outputEntry);
    return 0;
```

Even if it compiles, I haven't tried anything to see if it works afterwards, since I am looking into compiling this to actually get into dotnet...

**Comment by sixpindin** [comment]
@LaurentTreguier thanks for that, however there are a few compiler errors on my system.

For anybody else getting this issue. I've created a patch that compiles on my machine at-least. I've created an asp.net project and spun it up locally without issues. I'm not a c++ guy so the patch is probably flawed but it at-least allowed me to compile and get started with asp.net on my linux machine.
I've attached it here in-case anybody finds it useful.
[corefix-readdir.patch.txt](https://github.com/dotnet/corefx/files/417437/corefix-readdir.patch.txt)
